File: C++/src/Queue.cpp
#include "../include/Queue.h"
#include <stdexcept>

Queue::Queue() : front(nullptr), back(nullptr), size(0) {}

Queue::Queue(const Queue &other) : front(nullptr), back(nullptr), size(0) {
  Node *current = other.front;
  while (current != nullptr) {
    enqueue(current->data);
    current = current->next;
  }
}

Queue::~Queue() { clear(); }

Queue &Queue::operator=(const Queue &other) {
  if (this != &other) {
    clear();
    Node *current = other.front;
    while (current != nullptr) {
      enqueue(current->data);
      current = current->next;
    }
  }
  return *this;
}

void Queue::enqueue(const std::string &value) {
  Node *newNode = new Node(value);
  if (back == nullptr) {
    front = back = newNode;
  } else {
    back->next = newNode;
    back = newNode;
  }
  size++;
}

std::string Queue::dequeue() {
  if (front == nullptr) {
    throw std::runtime_error("Queue is empty");
  }

  Node *temp = front;
  std::string value = front->data;
  front = front->next;

  if (front == nullptr) {
    back = nullptr;
  }

  delete temp;
  size--;
  return value;
}

std::string Queue::peek() const {
  if (front == nullptr) {
    throw std::runtime_error("Queue is empty");
  }
  return front->data;
}

void Queue::clear() {
  while (front != nullptr) {
    Node *temp = front;
    front = front->next;
    delete temp;
  }
  back = nullptr;
  size = 0;
}

void Queue::serialize(std::ofstream &out) const {
  out.write(reinterpret_cast<const char *>(&size), sizeof(size));
  Node *current = front;
  while (current != nullptr) {
    int len = current->data.length();
    out.write(reinterpret_cast<const char *>(&len), sizeof(len));
    out.write(current->data.c_str(), len);
    current = current->next;
  }
}

void Queue::deserialize(std::ifstream &in) {
  clear();
  int newSize;
  in.read(reinterpret_cast<char *>(&newSize), sizeof(newSize));

  for (int i = 0; i < newSize; i++) {
    int len;
    in.read(reinterpret_cast<char *>(&len), sizeof(len));
    char *buffer = new char[len + 1];
    in.read(buffer, len);
    buffer[len] = '\0';
    enqueue(std::string(buffer));
    delete[] buffer;
  }
}

File: go/structs/queue/queue.go
package queue

import (
	"encoding/json"
	"fmt"
)

// Node represents a queue node
type Node struct {
	Data string
	Next *Node
}

// Queue represents a FIFO data structure
type Queue struct {
	front *Node
	back  *Node
	size  int
}

// New creates a new Queue
func New() *Queue {
	return &Queue{
		front: nil,
		back:  nil,
		size:  0,
	}
}

// Enqueue adds an element to the back
func (q *Queue) Enqueue(value string) {
	newNode := &Node{
		Data: value,
		Next: nil,
	}

	if q.back == nil {
		q.front = newNode
		q.back = newNode
	} else {
		q.back.Next = newNode
		q.back = newNode
	}
	q.size++
}

// Dequeue removes and returns the front element
func (q *Queue) Dequeue() (string, error) {
	if q.front == nil {
		return "", fmt.Errorf("queue is empty")
	}

	value := q.front.Data
	q.front = q.front.Next

	if q.front == nil {
		q.back = nil
	}

	q.size--
	return value, nil
}

// Peek returns the front element without removing it
func (q *Queue) Peek() (string, error) {
	if q.front == nil {
		return "", fmt.Errorf("queue is empty")
	}
	return q.front.Data, nil
}

// Size returns the number of elements
func (q *Queue) Size() int {
	return q.size
}

// IsEmpty checks if queue is empty
func (q *Queue) IsEmpty() bool {
	return q.size == 0
}

// Clear removes all elements
func (q *Queue) Clear() {
	q.front = nil
	q.back = nil
	q.size = 0
}

// ToSlice returns all elements as slice (front to back)
func (q *Queue) ToSlice() []string {
	result := make([]string, 0, q.size)
	current := q.front
	for current != nil {
		result = append(result, current.Data)
		current = current.Next
	}
	return result
}

// MarshalJSON implements json.Marshaler
func (q *Queue) MarshalJSON() ([]byte, error) {
	return json.Marshal(q.ToSlice())
}

// UnmarshalJSON implements json.Unmarshaler
func (q *Queue) UnmarshalJSON(data []byte) error {
	var items []string
	if err := json.Unmarshal(data, &items); err != nil {
		return err
	}

	q.Clear()
	for _, item := range items {
		q.Enqueue(item)
	}
	return nil
}
